app/src/cose.h

```diff
@@ -27,30 +27,161 @@
+  {
+    UnprotectedHeader parsed;
+    // Adapted from parse_cose_header_parameters in t_cose_parameters.c.
+    // t_cose doesn't support custom header parameters yet.
+
+    QCBORError qcbor_result;
+
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(
+      &ctx, cbor::from_bytes(cose_sign1), QCBOR_DECODE_MODE_NORMAL);
+    QCBORDecode_EnterArray(&ctx, nullptr);
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError("Failed to parse COSE_Sign1 outer array");
+    }
+
+    uint64_t tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
+    if (tag != CBOR_TAG_COSE_SIGN1)
+    {
+      throw COSEDecodeError("COSE_Sign1 is not tagged");
+    }
+    QCBORItem coseItem;
+    QCBORDecode_GetNext(&ctx, &coseItem);
+
+    struct q_useful_buf_c unprotected_parameters;
+    QCBORDecode_EnterMap(&ctx, NULL);
+
+    enum
+    {
+      ALG_INDEX,
+      X5CHAIN_INDEX,
+      END_INDEX,
+    };
+    QCBORItem header_items[END_INDEX + 1];
+    header_items[X5CHAIN_INDEX].label.int64 = COSE_HEADER_PARAM_X5CHAIN;
+    header_items[X5CHAIN_INDEX].uLabelType = QCBOR_TYPE_INT64;
+    header_items[X5CHAIN_INDEX].uDataType = QCBOR_TYPE_ANY;
+    header_items[END_INDEX].uLabelType = QCBOR_TYPE_NONE;
+
+    QCBORDecode_GetItemsInMap(&ctx, header_items);
+
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      throw COSEDecodeError("Failed to decode unprotected header");
+    }
+    if (header_items[X5CHAIN_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.x5chain = std::vector<std::vector<uint8_t>>();
+
+      QCBORItem chainItem = header_items[X5CHAIN_INDEX];
+      if (chainItem.uDataType == QCBOR_TYPE_ARRAY)
+      {
+        int lenOfArray = chainItem.val.uCount;
+        if (lenOfArray == 0)
+        {
+          throw COSEDecodeError(
+            "x5chain array length was 0 in cose unprotected header.");
+        }
+        QCBORDecode_EnterArrayFromMapN(&ctx, COSE_HEADER_PARAM_X5CHAIN);
+        for (int i = 0; i < lenOfArray; i++)
+        {
+          QCBORDecode_GetNext(&ctx, &chainItem);
+          if (chainItem.uDataType == QCBOR_TYPE_BYTE_STRING)
+          {
+            parsed.x5chain->push_back(cbor::as_vector(chainItem.val.string));
+          }
+          else
+          {
+            throw COSEDecodeError(
+              "Next item in chain was not of type qcbor byte string.");
+          }
+        }
+        QCBORDecode_ExitArray(&ctx);
+      }
+      else if (chainItem.uDataType == QCBOR_TYPE_BYTE_STRING)
+      {
+        parsed.x5chain->push_back(cbor::as_vector(chainItem.val.string));
+      }
+      else
+      {
+        CCF_APP_INFO("x5c type", chainItem.uDataType);
+        throw COSEDecodeError(
+          "Value type of x5chain in unprotected header is not array or byte "
+          "string.");
+      }
+    }
+    QCBORDecode_ExitMap(&ctx);
+    QCBORDecode_ExitBstrWrapped(&ctx);
```

Comment:
This seems wrong. There's no bstr wrapping here.

Also we need to check for errors on the context by calling `QCBORDecode_Finish()` (this is also missing from `decode_protected_header` actually).